File: server/internal/auth/middleware.go
package auth

import (
	"context"
	"crypto/subtle"
	"fmt"
	"net/http"
	"slices"
	"strings"

	u "tallyo/internal/utils"
)

type contextKey string

type bearerIdentity struct {
	Subject  string
	Scopes   []string
	Timezone string
}

const (
	subjectKey contextKey = "auth_subject"
	scopesKey  contextKey = "auth_scopes"
)

func Subject(ctx context.Context) (string, bool) {
	subject, ok := ctx.Value(subjectKey).(string)
	return subject, ok
}

func ContextWithSubject(ctx context.Context, subject string) context.Context {
	return context.WithValue(ctx, subjectKey, subject)
}

func Scopes(ctx context.Context) []string {
	scopes, _ := ctx.Value(scopesKey).([]string)
	return scopes
}

func HasScope(ctx context.Context, scope string) bool {
	return slices.Contains(Scopes(ctx), scope)
}

func ContextWithScopes(ctx context.Context, scopes []string) context.Context {
	return context.WithValue(ctx, scopesKey, scopes)
}

func (s *Service) Protect(next http.Handler) http.Handler {
	return s.protect(next, "/.well-known/oauth-protected-resource")
}

func (s *Service) ProtectMCP(next http.Handler) http.Handler {
	return s.protect(next, "/.well-known/oauth-protected-resource/mcp")
}

func (s *Service) RequireOAuth(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !s.OAuthEnabled() {
			http.NotFound(w, r)
			return
		}
		next.ServeHTTP(w, r)
	})
}

func (s *Service) protect(next http.Handler, resourceMetadataPath string) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		cfg := s.config()
		if cfg.DisableAllAuth {
			ctx := ContextWithScopes(r.Context(), AllScopes)
			ctx = u.ContextWithTimezone(ctx, s.Timezone())
			next.ServeHTTP(w, r.WithContext(ctx))
			return
		}
		if identity, ok := s.bearerAuth(r); ok {
			ctx := ContextWithSubject(r.Context(), identity.Subject)
			ctx = ContextWithScopes(ctx, identity.Scopes)
			ctx = u.ContextWithTimezone(ctx, identity.Timezone)
			next.ServeHTTP(w, r.WithContext(ctx))
			return
		}
		if cfg.MasterPassword != "" && subtle.ConstantTimeCompare([]byte(r.Header.Get("X-API-Key")), []byte(cfg.MasterPassword)) == 1 {
			ctx := ContextWithScopes(r.Context(), AllScopes)
			ctx = u.ContextWithTimezone(ctx, s.Timezone())
			next.ServeHTTP(w, r.WithContext(ctx))
			return
		}
		w.Header().Set("WWW-Authenticate", fmt.Sprintf(`Bearer resource_metadata="%s"`, cfg.IssuerURL+resourceMetadataPath))
		http.Error(w, "unauthorized", http.StatusUnauthorized)
	})
}

func (s *Service) bearerAuth(r *http.Request) (bearerIdentity, bool) {
	parts := strings.Fields(r.Header.Get("Authorization"))
	if len(parts) != 2 || !strings.EqualFold(parts[0], "Bearer") {
		return bearerIdentity{}, false
	}
	claims, err := s.verifyAccessToken(parts[1])
	if err != nil {
		return bearerIdentity{}, false
	}
	timezone := claims.Locale.Timezone
	if timezone == "" {
		timezone = s.Timezone()
	}
	return bearerIdentity{Subject: claims.Subject, Scopes: strings.Fields(claims.Scope), Timezone: timezone}, true
}

func (s *Service) DevCORS(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		origin := strings.TrimRight(r.Header.Get("Origin"), "/")
		if s.devOriginAllowed(origin) {
			w.Header().Set("Access-Control-Allow-Origin", origin)
			w.Header().Set("Access-Control-Expose-Headers", "Mcp-Session-Id")
			w.Header().Set("Vary", "Origin")
			w.Header().Set("Access-Control-Allow-Headers", "Authorization, Content-Type, X-API-Key")
			w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PATCH, DELETE, OPTIONS")
		}
		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusNoContent)
			return
		}
		next.ServeHTTP(w, r)
	})
}

File: server-rs/src/auth/middleware.rs
use std::sync::Arc;

use axum::{
    body::Body,
    extract::{Request, State},
    http::{HeaderValue, Method, StatusCode},
    middleware::Next,
    response::{IntoResponse, Response},
};

use super::{ALL_SCOPES, Identity, Service, roles::parse_scopes};

pub async fn protect(State(service): State<Arc<Service>>, request: Request, next: Next) -> Response {
    protect_resource(service, request, next, "/.well-known/oauth-protected-resource").await
}
pub async fn protect_mcp(State(service): State<Arc<Service>>, request: Request, next: Next) -> Response {
    protect_resource(service, request, next, "/.well-known/oauth-protected-resource/mcp").await
}

async fn protect_resource(service: Arc<Service>, mut request: Request, next: Next, metadata_path: &str) -> Response {
    let bearer_identity = bearer_token(&request).and_then(|token| {
        service.verify_access_token(token).ok().map(|claims| Identity {
            subject: Some(claims.sub),
            scopes: parse_scopes(&claims.scope),
            timezone: if claims.locale.timezone.is_empty() { service.timezone() } else { claims.locale.timezone },
        })
    });
    let master_password = service.master_password().and_then(|password| {
        request
            .headers()
            .get("X-API-Key")
            .is_some_and(|value| subtle::ConstantTimeEq::ct_eq(value.as_bytes(), password.as_str().as_bytes()).into())
            .then_some(password)
    });
    let identity = match (master_password, bearer_identity) {
        _ if service.disable_all_auth() => Some(Identity {
            subject: None,
            scopes: ALL_SCOPES.to_vec(),
            timezone: service.timezone(),
        }),
        (_, Some(identity)) => Some(identity),
        (Some(_), _) => Some(Identity {
            subject: None,
            scopes: ALL_SCOPES.to_vec(),
            timezone: service.timezone(),
        }),
        _ => None,
    };
    if let Some(identity) = identity {
        request.extensions_mut().insert(identity);
        return next.run(request).await;
    }
    let mut response = Response::new(Body::from("unauthorized"));
    *response.status_mut() = StatusCode::UNAUTHORIZED;
    let header = format!("Bearer resource_metadata=\"{}{}\"", service.issuer_url(), metadata_path);
    response
        .headers_mut()
        .insert("WWW-Authenticate", HeaderValue::from_str(&header).unwrap());
    response
}

pub async fn require_oauth(State(service): State<Arc<Service>>, request: Request, next: Next) -> Response {
    if service.oauth_enabled() { next.run(request).await } else { StatusCode::NOT_FOUND.into_response() }
}

pub async fn dev_cors(State(service): State<Arc<Service>>, request: Request, next: Next) -> Response {
    let origin = request
        .headers()
        .get("Origin")
        .and_then(|value| value.to_str().ok())
        .map(|origin| origin.trim_end_matches('/').to_owned());
    let mut response = if request.method() == Method::OPTIONS {
        StatusCode::NO_CONTENT.into_response()
    } else {
        next.run(request).await
    };
    if let Some(origin) = origin.filter(|origin| service.dev_origin_allowed(origin)) {
        let headers = response.headers_mut();
        headers.insert("Access-Control-Allow-Origin", HeaderValue::from_str(&origin).unwrap());
        headers.insert(
            "Access-Control-Expose-Headers",
            HeaderValue::from_static("Mcp-Session-Id"),
        );
        headers.insert("Vary", HeaderValue::from_static("Origin"));
        headers.insert(
            "Access-Control-Allow-Headers",
            HeaderValue::from_static("Authorization, Content-Type, X-API-Key"),
        );
        headers.insert(
            "Access-Control-Allow-Methods",
            HeaderValue::from_static("GET, POST, PATCH, DELETE, OPTIONS"),
        );
    }
    response
}

fn bearer_token(request: &Request) -> Option<&str> {
    let values = request
        .headers()
        .get("Authorization")?
        .to_str()
        .ok()?
        .split_whitespace()
        .collect::<Vec<_>>();
    (values.len() == 2 && values[0].eq_ignore_ascii_case("bearer")).then_some(values[1])
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use axum::{
        Router,
        body::to_bytes,
        http::{Request, StatusCode},
        middleware,
        routing::get,
    };
    use tower::ServiceExt;

    use super::{dev_cors, protect, protect_mcp};
    use crate::{
        auth::{AuthSettings, Config, Scope, Service},
        database::dbtest,
        middleware::client_ip::ClientIpResolver,
    };

    async fn service() -> Arc<Service> {
        let config = Config::new(
            AuthSettings {
                issuer_url: "https://tallyo.test".to_owned(),
                oauth_enabled: true,
                frontend_redirect_uris: vec!["https://web.test/callback".to_owned()],
                dev_cors_allowed_origins: vec!["https://web.test/".to_owned()],
                ..Default::default()
            },
            ClientIpResolver::new(&[]).unwrap(),
        );
        Arc::new(Service::new(config, dbtest::open().await.unwrap()).await.unwrap())
    }

    #[tokio::test]
    async fn protects_bearer_requests_and_advertises_mcp_metadata() {
        let service = service().await;
        let token = service
            .mint_access_token("person@example.com", &[Scope::ReadAccounts], "UTC")
            .unwrap();
        let router = Router::new()
            .route("/", get(|| async { "ok" }))
            .layer(middleware::from_fn_with_state(Arc::clone(&service), protect));
        let response = router
            .oneshot(
                Request::builder()
                    .uri("/")
                    .header("Authorization", format!("Bearer {token}"))
                    .body(axum::body::Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(response.status(), StatusCode::OK);

        let router = Router::new()
            .route("/", get(|| async { "ok" }))
            .layer(middleware::from_fn_with_state(service, protect_mcp));
        let response = router
            .oneshot(Request::builder().uri("/").body(axum::body::Body::empty()).unwrap())
            .await
            .unwrap();
        assert_eq!(response.status(), StatusCode::UNAUTHORIZED);
        assert_eq!(
            response.headers()["WWW-Authenticate"],
            "Bearer resource_metadata=\"https://tallyo.test/.well-known/oauth-protected-resource/mcp\""
        );
        assert_eq!(to_bytes(response.into_body(), 1024).await.unwrap(), "unauthorized");
    }

    #[tokio::test]
    async fn applies_development_cors_only_to_allowed_origins() {
        let service = service().await;
        let router = Router::new()
            .route("/", get(|| async { "ok" }))
            .layer(middleware::from_fn_with_state(service, dev_cors));
        let response = router
            .clone()
            .oneshot(
                Request::builder()
                    .uri("/")
                    .header("Origin", "https://web.test/")
                    .body(axum::body::Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(response.status(), StatusCode::OK);
        assert_eq!(response.headers()["Access-Control-Allow-Origin"], "https://web.test");
        for header in [
            "Access-Control-Allow-Origin",
            "Access-Control-Expose-Headers",
            "Vary",
            "Access-Control-Allow-Headers",
            "Access-Control-Allow-Methods",
        ] {
            assert!(response.headers().contains_key(header));
        }

        let response = router
            .oneshot(
                Request::builder()
                    .uri("/")
                    .header("Origin", "https://other.test")
                    .body(axum::body::Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(response.status(), StatusCode::OK);
        assert!(response.headers().get("Access-Control-Allow-Origin").is_none());
    }
}
